pandas/_libs/custom_window.py

```diff
@@ -60,3 +56,295 @@ def get_window_bounds(
+
+        end_s = np.arange(win, dtype='int64') + 1
+        end_e = start_e + win
+        self.end = np.concatenate([end_s, end_e])
+        self.win = win
+
+
+class VariableWindowIndexer(WindowIndexer):
+    """
+    create a variable length window indexer object
+    that has start & end, that point to offsets in
+    the index object; these are defined based on the win
+    arguments
+
+    Parameters
+    ----------
+    values: ndarray
+        values data array
+    win: int64_t
+        window size
+    minp: int64_t
+        min number of obs in a window to consider non-NaN
+    index: ndarray
+        index of the values
+    left_closed: bint
+        left endpoint closedness
+        True if the left endpoint is closed, False if open
+    right_closed: bint
+        right endpoint closedness
+        True if the right endpoint is closed, False if open
+    floor: optional
+        unit for flooring the unit
+    """
+    def __init__(self,
+                 values: np.ndarray,
+                 win: int,
+                 minp: int,
+                 left_closed: bool,
+                 right_closed: bool,
+                 index: np.ndarray,
+                 floor=None):
+
+        self.is_variable = 1
+        self.N = len(index)
+
+        self.start = np.empty(self.N, dtype='int64')
+        self.start.fill(-1)
+
+        self.end = np.empty(self.N, dtype='int64')
+        self.end.fill(-1)
+
+        self.build(index, win, left_closed, right_closed)
+
+        # max window size
+        self.win = (self.end - self.start).max()
+
+    def build(self,
+              index: np.ndarray,
+              win: int,
+              left_closed: bool,
+              right_closed: bool):
+
+        start = self.start
+        end = self.end
+        N = self.N
+
+        start[0] = 0
+
+        # right endpoint is closed
+        if right_closed:
+            end[0] = 1
+        # right endpoint is open
+        else:
+            end[0] = 0
+
+        # start is start of slice interval (including)
+        # end is end of slice interval (not including)
+        for i in range(1, N):
+            end_bound = index[i]
+            start_bound = index[i] - win
+
+            # left endpoint is closed
+            if left_closed:
+                start_bound -= 1
+
+            # advance the start bound until we are
+            # within the constraint
+            start[i] = i
+            for j in range(start[i - 1], i):
+                if index[j] > start_bound:
+                    start[i] = j
+                    break
+
+            # end bound is previous end
+            # or current index
+            if index[end[i - 1]] <= end_bound:
+                end[i] = i + 1
+            else:
+                end[i] = end[i - 1]
+
+            # right endpoint is open
+            if not right_closed:
+                end[i] -= 1
+
+
+def get_window_indexer(values: np.ndarray, win: int, index: Optional[np.ndarray],
```

Comment:
why don't we decouple this, meaning the construction of the indexer and the calling of .get_window_bounds().

Then you don't need the custom_indexer arg / mock things.

The CustomIndexer just follows the interfae of the Fixed/Variable Indexer and then *inside* the code (of window.py) we can call .get_window_bounds()

Reply:
I think we're on the same page; I was aiming for just one `.get_window_bounds` call ideally having `Fixed`/`Variable` Indexer subclass the `BaseIndexer`